### ISSUE TITLE
[lld] location counter alignment not affecting . when assigning an output section address and a MEMORY region.

### ISSUE THREAD

**Issue by smithp35** [llvm/llvm-project#112919]
This example is derived from an internal linker script used with ld.bfd that brought up an unexpected warning with LLD. There are ways to work around it. Raising to log the discrepancy.

When assigning an OutputSection an address an explicit address and a MEMORY region, my reading of the ld.bfd manual is that the explicit address takes precedence over the location in the MEMORY region (https://sourceware.org/binutils/docs/ld/Output-Section-Address.html)

It looks like when assigning the OutputSection the value of the location counter as the address, lld uses the location counter from the MEMORY region and not the global location counter. For example:
```
        .section s01, "ax", %progbits
        .byte 1

        .section s02, "ax", %progbits
        .balign 4
        .word 1
```
With linker script:
```
MEMORY
{
  FLASH (rx) : ORIGIN = 0x1000, LENGTH = 0x100
  RAM (rw) : ORIGIN = 0x2000, LENGTH = 0x1000
}

SECTIONS
{
 /* s01 is 1 byte */
  s01 : { *(s01) } > FLASH
 /* . == 1 */
  . = ALIGN(4);
 /* . == 4 expectation that s02 gets 4-byte aligned address */
  s02 . : { *(s02) } > FLASH
}
```
However we get 
```
ld.lld: warning: address (0x1001) of section s02 is not a multiple of alignment (4)
```
With Map file:
```
             VMA              LMA     Size Align Out     In      Symbol
            1000             1000        1     1 s01
            1000             1000        1     1         align.o:(s01)
            1001             1001        3     1 . = ALIGN(4)
            1001             1001        5     4 s02
            1004             1004        2     4         align.o:(s02)
```
Whereas with ld.bfd we get
```
s01             0x0000000000001000        0x1
 *(s01)
 s01            0x0000000000001000        0x1 align.o
                0x0000000000001004                . = ALIGN (0x4)

s02             0x0000000000001004        0x2
 *(s02)
 s02            0x0000000000001004        0x2 align.o
```
If I move the ALIGN to inside the s01 OutputSection `{ *(s01); . = ALIGN(4); }`, we get similar results
```
            1000             1000        4     1 s01
            1000             1000        1     1         align.o:(s01)
            1001             1001        3     1         . = ALIGN(4)
            1004             1004        2     4 s02
            1004             1004        2     4         align.o:(s02)
```

My suspicion is that when using `.` in the context of Memory regions is using the location counter from the MEMORY region and not the global location counter.
